In [82]:
import pandas as pd

In [83]:
flights=pd.read_csv('flights.csv')
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute  
0    1545    EWR  IAH     227.0      1400   5.0    17.0  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  
3     725    JFK  BQN     183.0      1576   5.0    44.0  
4     461    LGA  ATL     116.0       762   5.0    54.0

1. How many total number of days does the flights table cover?

In [85]:
total_days = flights['day'].sum()
total_days

5291016

Ans : There are total 5291016 number of days in flight table.

In [86]:
airports = pd.read_csv('airports.csv')
airports.head()

IATA_CODE                              AIRPORT         CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [87]:
airports.rename(columns={'IATA_CODE':'origin'}, inplace=True)
airports

origin                              AIRPORT  \
0      ABE  Lehigh Valley International Airport   
1      ABI             Abilene Regional Airport   
2      ABQ    Albuquerque International Sunport   
3      ABR            Aberdeen Regional Airport   
4      ABY   Southwest Georgia Regional Airport   
..     ...                                  ...   
317    WRG                     Wrangell Airport   
318    WYS               Westerly State Airport   
319    XNA  Northwest Arkansas Regional Airport   
320    YAK                      Yakutat Airport   
321    YUM           Yuma International Airport   

                               CITY STATE COUNTRY  LATITUDE  LONGITUDE  
0                         Allentown    PA     USA  40.65236  -75.44040  
1                           Abilene    TX     USA  32.41132  -99.68190  
2                       Albuquerque    NM     USA  35.04022 -106.60919  
3                          Aberdeen    SD     USA  45.44906  -98.42183  
4                            Albany    GA     USA  31.53552  -84.19447  
..                              ...   ...     ...       ...        ...  
317                        Wrangell    AK     USA  56.48433 -132.36982  
318                West Yellowstone    MT     USA  44.68840 -111.11764  
319  Fayetteville/Springdale/Rogers    AR     USA  36.28187  -94.30681  
320                         Yakutat    AK     USA  59.50336 -139.66023  
321                            Yuma    AZ     USA  32.65658 -114.60597  

[322 rows x 7 columns]

In [88]:
flights['CITY'] = flights.origin.map(airports.set_index('origin')['CITY'])
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute      CITY  
0    1545    EWR  IAH     227.0      1400   5.0    17.0    Newark  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  New York  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  New York  
3     725    JFK  BQN     183.0      1576   5.0    44.0  New York  
4     461    LGA  ATL     116.0       762   5.0    54.0  New York

2. How many departure cities (not airports) does the flights database cover?

In [89]:
dep_cities = flights['CITY'].count()
dep_cities

336776

Ans : There are 336776 departure cities in flight table.

In [93]:
planes=pd.read_csv('planes.csv')
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

3. What is the relationship between flights and planes tables?

Ans : Many to One relationship between flights and planes tables.

4. Which airplane manufacturer incurred the most delays in the analysis period?

In [90]:
flights['manufacturer']= flights.tailnum.map(planes.set_index('tailnum')['manufacturer'])
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute      CITY manufacturer  
0    1545    EWR  IAH     227.0      1400   5.0    17.0    Newark       BOEING  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  New York       BOEING  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  New York       BOEING  
3     725    JFK  BQN     183.0      1576   5.0    44.0  New York       AIRBUS  
4     461    LGA  ATL     116.0       762   5.0    54.0  New York       BOEING

In [95]:
flights.groupby('manufacturer')['arr_delay'].count().reset_index(name='most_no_of_delays')\
                                 .sort_values(['most_no_of_delays'], ascending=False).head(1)

manufacturer  most_no_of_delays
9       BOEING              82283

Ans : Manufacturer Boeing incurred the most delays in the analysis period. 

5. Which are the two most connected cities?

In [94]:
flights.groupby('CITY')['flight'].count().reset_index(name='MOST_NO_OF_FLIGHTS')\
                                 .sort_values(['MOST_NO_OF_FLIGHTS'], ascending=False)

CITY  MOST_NO_OF_FLIGHTS
0  New York              215941
1    Newark              120835

Ans : Since New York and Newark has the most number of flights that means they are the 2 most connected         cities.